src/language/language.rs

```diff
@@ -1641,6 +1742,117 @@ impl egg::Analysis<Language> for MyAnalysis {
+                                    / strides.shape[1];
+                            }
+                            crate::language::RelayActivationLayout::NHWC => {
+                                // Sorry for the horrific indentation...
+                                access[1] =
+                                // The dimension plus padding
+                                    (((padding.shape[0] + access[1] + padding.shape[2])
+                                      // Get the number of spots where we could pool
+                                      - (pool_size.shape[0] - 1))
+                                     // Then calculate the spots we actually pool at
+                                     // using the stride
+                                     + strides.shape[0]
+                                     - 1)
+                                    / strides.shape[0];
+                                access[2] = (((padding.shape[1] + access[2] + padding.shape[3])
+                                              // Get the number of spots where we could pool
+                                              - (pool_size.shape[1] - 1))
+                                             // Then calculate the spots we actually pool at
+                                             // using the stride
+                                             + strides.shape[1]
+                                    - 1)
+                                    / strides.shape[1];
+                            }
+                        }
+
+                        MyAnalysisData::AccessPattern(access)
+                    }
+                    crate::language::RelayOperator::RelayUpSampling => {
+                        let mut access = match params[1..]
+                            .iter()
+                            .map(|id| &egraph[*id].data)
+                            .collect::<Vec<_>>()[..]
+                        {
+                            [MyAnalysisData::AccessPattern(a), MyAnalysisData::Literal(scale_h), MyAnalysisData::Literal(scale_w), MyAnalysisData::RelayActivationLayout(layout)] =>
+                            {
+                                assert_eq!(
+                                    layout.clone(),
+                                    crate::language::RelayActivationLayout::NCHW,
+                                    "upsampling only supports NCHW"
+                                );
+                                // let mut shape = array![a.shape[0], a.shape[1], scale_h.into() * shape[2], scale_w.into() * shape[w]];
+                                let mut shape = a.shape.clone();
+                                assert_eq!(scale_h.ndim(), 0);
+                                assert_eq!(scale_w.ndim(), 0);
+                                shape[2] = (scale_h.first().unwrap() * (shape[2] as f64)) as usize;
+                                shape[3] = (scale_w.first().unwrap() * (shape[3] as f64)) as usize;
```

Comment:
Does the right kind of rounding happen here? What happens if the floats are not integers?

Reply:
it should fail!

Reply:
I'm unsure about this. The documentation for it in Relay implies there's more complexity here: https://tvm.apache.org/docs/api/python/relay/nn.html#tvm.relay.nn.upsampling. I think they would have made the parameters integers, if they could only be integers. For now, can we just assert that the float values must be integers? And then in the future we can support fractional values if we need to.

Reply:
I looked in to the TVM implementation a bit more. See [this code in TOPI](https://github.com/apache/tvm/blob/da446af3b73a9b7354b89bb5792968cd705359d8/python/tvm/topi/nn/upsampling.py#L67). The scales are multiplied as-is (i.e., with the fractional part) with the size of the image to calculate the new size of the image. So I think they can be fractional!

Reply:
hmm it looks like right after that part of the code they calculate the new resized_shape https://github.com/apache/tvm/blob/da446af3b73a9b7354b89bb5792968cd705359d8/python/tvm/topi/nn/upsampling.py#L70

                simplify(topi.cast(te.round(scaled_h), data.shape[2].dtype)),
                simplify(topi.cast(te.round(scaled_w), data.shape[3].dtype)),

which looks like it rounds the scaled numbers to the nearest integer if I am reading this right?

Reply:
I guess the question is: does `te.round()` round the same as `as usize`? If that's the case, then yeah, your code is right.

Can you use [round](https://doc.rust-lang.org/std/primitive.f64.html#method.round) instead of just `as usize`? That will bring us a little closer to the behavior, I think. Also, can you add a TODO saying that the rounding may not be exactly the same as TVM's?

Honestly, there's no good answer if we want to be exact here, and I'm just going to bank on it not being a problem!

Reply:
added round!